mono/utils/mono-threads.h

```diff
@@ -751,4 +797,84 @@ typedef void (*background_job_cb)(void);
 void mono_threads_schedule_background_job (background_job_cb cb);
 #endif
 
+#ifdef MONO_THREADS_NEED_THREAD_INFO_TYPE_WRAPPERS
+#undef MONO_THREADS_NEED_THREAD_INFO_TYPE_WRAPPERS
+
+// This scheme for dealing with THREAD_INFO_TYPE differs
+// from the old scheme, in that it has function prototypes match function implementations.
+// It does break taking address of functions ("can't take address of macros").
+// FIXME In C++ future, can replace this hack with derivation from MonoThreadInfo.
+// That will require changing all the explicit accesses to the "named base".
+// SgenThreadInfo * info;
+// info->base.foo => info-foo, and can be only one foo between base and derived.
+// This at least solves the input casts, which is most of the problem.
+
+// Output casts.
+#define mono_thread_info_attach(info)		((THREAD_INFO_TYPE*)mono_thread_info_attach ())
+#define mono_thread_info_current()		((THREAD_INFO_TYPE*)mono_thread_info_current ())
+#define mono_thread_info_current_unchecked()	((THREAD_INFO_TYPE*)mono_thread_info_current_unchecked ())
+#define mono_thread_info_lookup(id) 		((THREAD_INFO_TYPE*)mono_thread_info_lookup (id))
```

Comment:
These macros can all be functions, if that is more palatable.
It is obvious for the "input" form, just overload.

The "output" form goes like:
```
static inline THREAD_INFO_TYPE*
mono_thread_info_lookup (id, THREAD_INFO_TYPE *  = 0 )
{
	return (THREAD_INFO_TYPE*)mono_thread_info_lookup (id, (MonoThreadInfo*)0);
}
```

Notice the defaulted parameter.
The main functions would get this too though, otherwise is ambiguous.